In [4]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

In [5]:
crime_df=pd.read_csv(r'Part_1_Crime_Data.csv', low_memory=False)
crime_df.head()

X          Y  RowID  CCNumber           CrimeDateTime CrimeCode  \
0 -76.609534  39.318181      1  23D05796  2023/04/15 11:30:00+00        6F   
1 -76.530961  39.310944      2  23D04902  2023/04/15 21:50:00+00        6C   
2 -76.674519  39.313038      3  23D04906  2023/04/15 17:00:00+00        6D   
3 -76.664090  39.337523      4  23D04912  2023/04/15 21:15:00+00       3AF   
4 -76.661570  39.296418      5  23D03209  2023/04/11 09:25:00+00        4E   

         Description Inside_Outside            Weapon   Post  ...  \
0            LARCENY         Inside               NaN  513.0  ...   
1            LARCENY         Inside               NaN  441.0  ...   
2  LARCENY FROM AUTO         Inside               NaN  812.0  ...   
3            ROBBERY         Inside           HANDGUN  614.0  ...   
4     COMMON ASSAULT        Outside  PERSONAL_WEAPONS  721.0  ...   

                Ethnicity               Location Old_District New_District  \
0  NOT_HISPANIC_OR_LATINO    2500 GREENMOUNT AVE     NORTHERN          NaN   
1                     NaN           1500 62ND ST    NORTHEAST          NaN   
2  NOT_HISPANIC_OR_LATINO         3400 ELGIN AVE    SOUTHWEST          NaN   
3  NOT_HISPANIC_OR_LATINO  2600 PARK HEIGHTS TER    NORTHWEST          NaN   
4  NOT_HISPANIC_OR_LATINO       700 ASHBURTON ST      WESTERN          NaN   

              Neighborhood   Latitude  Longitude  \
0                  HARWOOD  39.318181 -76.609534   
1  PULASKI INDUSTRIAL AREA  39.310944 -76.530961   
2              MOUNT HOLLY  39.313038 -76.674519   
3     CENTRAL PARK HEIGHTS  39.337523 -76.664090   
4                   MOSHER  39.296418 -76.661570   

                                GeoLocation           PremiseType  \
0  (39.318181000000003,-76.609533999999996)                STREET   
1  (39.310943999999999,-76.530961000000005)           GAS STATION   
2  (39.313037999999999,-76.674519000000004)                STREET   
3  (39.337522999999997,-76.664090000000002)                STREET   
4  (39.296418000000003,-76.661569999999998)  DRUG STORE / MED  BL   

   Total_Incidents  
0                1  
1                1  
2                1  
3                1  
4                1  

[5 rows x 23 columns]

In [6]:
crime_df.shape

(603894, 23)

In [7]:
crime_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 603894 entries, 0 to 603893
Data columns (total 23 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   X                602888 non-null  float64
 1   Y                602888 non-null  float64
 2   RowID            603894 non-null  int64  
 3   CCNumber         603894 non-null  object 
 4   CrimeDateTime    603894 non-null  object 
 5   CrimeCode        603894 non-null  object 
 6   Description      603894 non-null  object 
 7   Inside_Outside   550890 non-null  object 
 8   Weapon           146702 non-null  object 
 9   Post             595977 non-null  float64
 10  Gender           507584 non-null  object 
 11  Age              486543 non-null  float64
 12  Race             578124 non-null  object 
 13  Ethnicity        76269 non-null   object 
 14  Location         600484 non-null  object 
 15  Old_District     567502 non-null  object 
 16  New_District     28478 non-null   obje

In [8]:
crime_df.duplicated().sum()

0

In [9]:
crime_df.isnull().sum()

X                    1006
Y                    1006
RowID                   0
CCNumber                0
CrimeDateTime           0
CrimeCode               0
Description             0
Inside_Outside      53004
Weapon             457192
Post                 7917
Gender              96310
Age                117351
Race                25770
Ethnicity          527625
Location             3410
Old_District        36392
New_District       575416
Neighborhood         8563
Latitude             1006
Longitude            1006
GeoLocation             0
PremiseType         53340
Total_Incidents         0
dtype: int64

In [10]:
df=crime_df[['CrimeDateTime', 'Description', 'Inside_Outside', 'Weapon','Gender','Age','Race', 'Old_District', 'Neighborhood','Latitude','Longitude','PremiseType','Total_Incidents']]

In [11]:
df.shape

(603894, 13)

In [12]:
df.isnull().sum()

CrimeDateTime           0
Description             0
Inside_Outside      53004
Weapon             457192
Gender              96310
Age                117351
Race                25770
Old_District        36392
Neighborhood         8563
Latitude             1006
Longitude            1006
PremiseType         53340
Total_Incidents         0
dtype: int64

#### Inside_Outside

In [13]:
df.Inside_Outside.unique()

array(['Inside', 'Outside', nan, 'O', 'I'], dtype=object)

In [14]:
df['Inside_Outside']=df['Inside_Outside'].replace({'O': 'Outside','I': 'Inside'})

In [15]:
df.Inside_Outside.value_counts()

Inside_Outside
Outside    284058
Inside     266832
Name: count, dtype: int64

In [16]:
df = df.dropna(subset=['Inside_Outside'])

#### Weapon

In [17]:
df.Weapon.unique()

array([nan, 'HANDGUN', 'PERSONAL_WEAPONS', 'AUTOMATIC_HANDGUN',
       'BLUNT_OBJECT', 'FIREARM', 'KNIFE_CUTTING_INSTRUMENT',
       'MOTOR_VEHICLE_VESSEL', 'KNIFE', 'OTHER', 'HANDS', 'RIFLE', 'FIRE',
       'UNKNOWN', 'ASPHYXIATION', 'AUTOMATIC_FIREARM', 'OTHER_FIREARM',
       'POISON', 'SHOTGUN', 'AUTOMATIC_RIFLE', 'FIRE_INCENDIARY_DEVICE',
       'EXPLOSIVES', 'AUTOMATIC_SHOTGUN',
       'DRUGS_NARCOTICS_SLEEPING_PILLS'], dtype=object)

In [18]:
df.Weapon.isnull().sum()

414288

In [19]:
df['Weapon'].fillna('UNKNOWN', inplace=True)

In [20]:
df.Weapon.isnull().sum()

0

In [21]:
df.Weapon.value_counts()

Weapon
UNKNOWN                           415327
FIREARM                            42639
PERSONAL_WEAPONS                   29542
OTHER                              26995
KNIFE                              15156
HANDGUN                             6355
HANDS                               6201
KNIFE_CUTTING_INSTRUMENT            3246
BLUNT_OBJECT                        2177
FIRE                                2040
MOTOR_VEHICLE_VESSEL                 458
AUTOMATIC_HANDGUN                    128
ASPHYXIATION                         126
OTHER_FIREARM                        115
RIFLE                                112
SHOTGUN                               80
FIRE_INCENDIARY_DEVICE                74
AUTOMATIC_FIREARM                     61
POISON                                29
AUTOMATIC_RIFLE                       17
DRUGS_NARCOTICS_SLEEPING_PILLS         5
EXPLOSIVES                             4
AUTOMATIC_SHOTGUN                      3
Name: count, dtype: int64

#### Gender

In [22]:
df.Gender.unique()

array(['F', nan, 'M', 'Male', 'Female', 'U', 'FB', 'Transgende', 'B', ',',
       'N', 'D', 'FW', 'FU', 'W', 'FF', 'T', 'M\\', '160', '042819', 'FM',
       'A', '165', '17', '77', 'MB', '8', '120'], dtype=object)

In [23]:
df['Gender']=df['Gender'].replace({'F': 'Female', 'M':'Male','U':'Other', 
                                                    'FB':'Other', 'Transgende':'Other', 'B':'Other', ',':'Other',
                                                    'N':'Other', 'D':'Other', 'FW':'Other', 'FU':'Other', 'W':'Other', 
                                                    'FF':'Other', 'T':'Other', 'M\\':'Other', '160':'Other', '042819':'Other', 
                                                    'FM':'Other','A':'Other', '165':'Other', '17':'Other', '77':'Other', 
                                                     'MB':'Other', '8':'Other', '120':'Other'})

In [24]:
df.Gender.unique()

array(['Female', nan, 'Male', 'Other'], dtype=object)

In [25]:
df.Gender.isnull().sum()

83612

In [26]:
df.Gender.value_counts()

Gender
Female    234181
Male      221806
Other      11291
Name: count, dtype: int64

In [27]:
df['Gender'].fillna('UNKNOWN', inplace=True)

In [28]:
df.Gender.value_counts()

Gender
Female     234181
Male       221806
UNKNOWN     83612
Other       11291
Name: count, dtype: int64

In [29]:
df.shape

(550890, 13)

#### Age

In [30]:
(df.Age<0).any()

True

In [31]:
df.Age.isnull().sum()

101900

In [32]:
non_valid_age = (df['Age'] < 0) | (df['Age'] > 120)
non_valid_age.sum()

156

There are 101900 null values and also 156 invalid values for the Age column in the dataset. So, let us replace these values with mean Age.

In [33]:
#Repalce null values with the mean age
df['Age'].fillna(df[df['Age'] >= 1]['Age'].mean(), inplace=True)

In [34]:
df.Age.isnull().sum()

0

In [35]:
mean_age = df[df['Age'] > 1]['Age'].mean()
mean_age

38.153726268572534

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 550890 entries, 0 to 603893
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   CrimeDateTime    550890 non-null  object 
 1   Description      550890 non-null  object 
 2   Inside_Outside   550890 non-null  object 
 3   Weapon           550890 non-null  object 
 4   Gender           550890 non-null  object 
 5   Age              550890 non-null  float64
 6   Race             527227 non-null  object 
 7   Old_District     518562 non-null  object 
 8   Neighborhood     543220 non-null  object 
 9   Latitude         549953 non-null  float64
 10  Longitude        549953 non-null  float64
 11  PremiseType      550554 non-null  object 
 12  Total_Incidents  550890 non-null  int64  
dtypes: float64(3), int64(1), object(9)
memory usage: 58.8+ MB


In [37]:
df[~non_valid_age].info()

<class 'pandas.core.frame.DataFrame'>
Index: 550734 entries, 0 to 603893
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   CrimeDateTime    550734 non-null  object 
 1   Description      550734 non-null  object 
 2   Inside_Outside   550734 non-null  object 
 3   Weapon           550734 non-null  object 
 4   Gender           550734 non-null  object 
 5   Age              550734 non-null  float64
 6   Race             527076 non-null  object 
 7   Old_District     518408 non-null  object 
 8   Neighborhood     543066 non-null  object 
 9   Latitude         549797 non-null  float64
 10  Longitude        549797 non-null  float64
 11  PremiseType      550403 non-null  object 
 12  Total_Incidents  550734 non-null  int64  
dtypes: float64(3), int64(1), object(9)
memory usage: 58.8+ MB


In [38]:
df=df[~non_valid_age]
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 550734 entries, 0 to 603893
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   CrimeDateTime    550734 non-null  object 
 1   Description      550734 non-null  object 
 2   Inside_Outside   550734 non-null  object 
 3   Weapon           550734 non-null  object 
 4   Gender           550734 non-null  object 
 5   Age              550734 non-null  float64
 6   Race             527076 non-null  object 
 7   Old_District     518408 non-null  object 
 8   Neighborhood     543066 non-null  object 
 9   Latitude         549797 non-null  float64
 10  Longitude        549797 non-null  float64
 11  PremiseType      550403 non-null  object 
 12  Total_Incidents  550734 non-null  int64  
dtypes: float64(3), int64(1), object(9)
memory usage: 58.8+ MB


In [39]:
df.isnull().sum()

CrimeDateTime          0
Description            0
Inside_Outside         0
Weapon                 0
Gender                 0
Age                    0
Race               23658
Old_District       32326
Neighborhood        7668
Latitude             937
Longitude            937
PremiseType          331
Total_Incidents        0
dtype: int64

In [40]:
df.dtypes

CrimeDateTime       object
Description         object
Inside_Outside      object
Weapon              object
Gender              object
Age                float64
Race                object
Old_District        object
Neighborhood        object
Latitude           float64
Longitude          float64
PremiseType         object
Total_Incidents      int64
dtype: object

In [41]:
df['Age']=df['Age'].astype(int)

In [42]:
df.dtypes

CrimeDateTime       object
Description         object
Inside_Outside      object
Weapon              object
Gender              object
Age                  int32
Race                object
Old_District        object
Neighborhood        object
Latitude           float64
Longitude          float64
PremiseType         object
Total_Incidents      int64
dtype: object

In [43]:
df.shape

(550734, 13)

#### Race

In [44]:
df.Race.unique()

array(['WHITE', nan, 'BLACK_OR_AFRICAN_AMERICAN', 'UNKNOWN',
       'AMERICAN_INDIAN_OR_ALASKA_NATIVE', 'ASIAN',
       'NATIVE_HAWAIIAN_OR_OTHER_PACIFIC_ISLANDER'], dtype=object)

In [45]:
df.Race.value_counts()

Race
BLACK_OR_AFRICAN_AMERICAN                    295387
WHITE                                        115488
UNKNOWN                                      113184
ASIAN                                          1793
AMERICAN_INDIAN_OR_ALASKA_NATIVE                804
NATIVE_HAWAIIAN_OR_OTHER_PACIFIC_ISLANDER       420
Name: count, dtype: int64

In [46]:
df['Race'].fillna('UNKNOWN', inplace=True)

In [47]:
df.Race.value_counts()

Race
BLACK_OR_AFRICAN_AMERICAN                    295387
UNKNOWN                                      136842
WHITE                                        115488
ASIAN                                          1793
AMERICAN_INDIAN_OR_ALASKA_NATIVE                804
NATIVE_HAWAIIAN_OR_OTHER_PACIFIC_ISLANDER       420
Name: count, dtype: int64

#### Old_District

In [48]:
df.Old_District.unique()

array(['NORTHERN', 'NORTHEAST', 'SOUTHWEST', 'NORTHWEST', 'WESTERN',
       'CENTRAL', 'SOUTHEAST', 'EASTERN', 'SOUTHERN', nan], dtype=object)

In [49]:
df.Old_District.value_counts()

Old_District
NORTHEAST    79362
SOUTHEAST    71196
CENTRAL      60535
SOUTHERN     59559
NORTHERN     55843
NORTHWEST    52787
SOUTHWEST    52045
EASTERN      45362
WESTERN      41719
Name: count, dtype: int64

In [50]:
df.Old_District.isnull().sum()

32326

In [51]:
#df = df.dropna(subset=['Old_District'])
df['Old_District'].fillna('UNKNOWN', inplace=True)

In [52]:
df.shape

(550734, 13)

#### Neighborhood

In [53]:
df.Neighborhood.unique()

array(['HARWOOD', 'PULASKI INDUSTRIAL AREA', 'MOUNT HOLLY',
       'CENTRAL PARK HEIGHTS', 'MOSHER', 'MID-TOWN BELVEDERE',
       'BALTIMORE HIGHLANDS', 'DOWNTOWN', 'OLDTOWN', 'CHERRY HILL',
       'YALE HEIGHTS', 'CEDMONT', 'LAKELAND', 'FRANKFORD', 'FELLS POINT',
       'ROSEMONT HOMEOWNERS/TENANTS', 'BOLTON HILL', 'WAVERLY',
       'IRVINGTON', 'CANTON', 'PATTERSON PLACE', 'MORRELL PARK',
       'MORAVIA-WALTHER', 'SANDTOWN-WINCHESTER', 'CANTON INDUSTRIAL AREA',
       'MADISON-EASTEND', 'PENROSE/FAYETTE STREET OUTREACH',
       'FRANKLIN SQUARE', 'WOODMERE', 'GWYNNS FALLS', 'WINSTON-GOVANS',
       'ORANGEVILLE INDUSTRIAL AREA', 'BURLEITH-LEIGHTON',
       'HAMILTON HILLS', 'EAST BALTIMORE MIDWAY', 'HANLON-LONGWOOD',
       'HUNTING RIDGE', 'GARWYN OAKS', 'VIOLETVILLE', "BUTCHER'S HILL",
       'PORT COVINGTON', 'CURTIS BAY', 'OLD GOUCHER', 'WINCHESTER',
       'MILLHILL', 'SETON HILL', 'LOCH RAVEN', 'WALTHERSON',
       'CARROLL - CAMDEN INDUSTRIAL AREA', 'NORTH HARFORD ROAD',
    

In [54]:
df.Neighborhood.isnull().sum()

7668

In [55]:
#df = df.dropna(subset=['Neighborhood'])
df['Neighborhood'].fillna('UNKNOWN', inplace=True)

In [56]:
df.shape

(550734, 13)

In [57]:
df.isnull().sum()

CrimeDateTime        0
Description          0
Inside_Outside       0
Weapon               0
Gender               0
Age                  0
Race                 0
Old_District         0
Neighborhood         0
Latitude           937
Longitude          937
PremiseType        331
Total_Incidents      0
dtype: int64

#### PremiseType

In [58]:
df.PremiseType.unique()

array(['STREET', 'GAS STATION', 'DRUG STORE / MED  BL', 'HOTEL/MOTEL',
       'OTHER/RESIDENTIAL', 'SCHOOL', 'CONVENIENCE STORE',
       'OFFICE BUILDING', 'SPECIALTY STORE', 'PUBLIC AREA', 'SHED/GARAGE',
       'RESTAURANT', 'TAVERN/NIGHT CLUB', 'OTHER - OUTSIDE     ',
       'STREET              ', 'APT/CONDO - OCCUPIED',
       'GAS STATION         ', 'ROW/TOWNHOUSE-OCC',
       'BUS/RAILROAD STATION', 'YARD', 'HOSP/NURS. HOME',
       'GARAGE ON PRIV. PROP', 'MARKET STALLS       ', 'ALLEY',
       'ROW/TOWNHOUSE-VAC', 'HOSPITAL', 'SCHOOL              ',
       'INNER HARBOR-S-KEY  ', 'CLOTHING/SHOE STORE ',
       'ATM MACHINES        ', 'RETAIL/SMALL BUSINES', 'PUBLIC BUILDING',
       'RELIGIOUS  INSTITUTI', 'LIQUOR STORE', 'OTHER - INSIDE      ',
       'PARKING LOT-OUTSIDE ', 'GROCERY/CONFECTION', 'DWELLING',
       'SHOPPING MALLS/CNTR ', 'PARK', 'VACANT BUILDING',
       'GROCERY/CONFECTION ', 'CARRY OUT', 'UNKNOWN', 'CONSTRUCTION SITE',
       'PORCH/DECK          ', 'TAVERN

In [59]:
df['PremiseType'] = df['PremiseType'].str.strip()

In [60]:
df.PremiseType.unique()

array(['STREET', 'GAS STATION', 'DRUG STORE / MED  BL', 'HOTEL/MOTEL',
       'OTHER/RESIDENTIAL', 'SCHOOL', 'CONVENIENCE STORE',
       'OFFICE BUILDING', 'SPECIALTY STORE', 'PUBLIC AREA', 'SHED/GARAGE',
       'RESTAURANT', 'TAVERN/NIGHT CLUB', 'OTHER - OUTSIDE',
       'APT/CONDO - OCCUPIED', 'ROW/TOWNHOUSE-OCC',
       'BUS/RAILROAD STATION', 'YARD', 'HOSP/NURS. HOME',
       'GARAGE ON PRIV. PROP', 'MARKET STALLS', 'ALLEY',
       'ROW/TOWNHOUSE-VAC', 'HOSPITAL', 'INNER HARBOR-S-KEY',
       'CLOTHING/SHOE STORE', 'ATM MACHINES', 'RETAIL/SMALL BUSINES',
       'PUBLIC BUILDING', 'RELIGIOUS  INSTITUTI', 'LIQUOR STORE',
       'OTHER - INSIDE', 'PARKING LOT-OUTSIDE', 'GROCERY/CONFECTION',
       'DWELLING', 'SHOPPING MALLS/CNTR', 'PARK', 'VACANT BUILDING',
       'CARRY OUT', 'UNKNOWN', 'CONSTRUCTION SITE', 'PORCH/DECK',
       'STADIUM', 'DEPARTMENT STORE', 'YARD/BUSINESS', 'BUS/AUTO',
       'TRACTOR TRAILER', 'LAUNDRY/CLEANRS/SHOE', 'VACANT LOT/WOODS',
       'PARKING LOT', 'BARB

In [61]:
df.PremiseType.value_counts()

PremiseType
STREET                  211341
ROW/TOWNHOUSE-OCC        95929
OTHER/RESIDENTIAL        42945
APT/CONDO - OCCUPIED     20734
PARKING LOT-OUTSIDE      17794
                         ...  
DAY CARE FACILITY            5
TRUCKING & TRANSFER          5
MANUFACTURING PLANT          4
BOX CARS/COMM'L TRLR         3
TRUCK DRIVERS-OIL-           2
Name: count, Length: 127, dtype: int64

In [62]:
(df['PremiseType']=='UNKNOWN').value_counts()

PremiseType
False    545467
True       5267
Name: count, dtype: int64

In [63]:
df.PremiseType.isnull().sum()

331

In [64]:
#df = df.dropna(subset=['PremiseType'])
df['PremiseType'].fillna('UNKNOWN', inplace=True)

In [65]:
df.PremiseType.isnull().sum()

0

#### Latitude and Longitude

In [66]:
df.Latitude.isnull().sum()

937

In [67]:
df = df.dropna(subset=['Latitude'])

In [68]:
#df.Longitude.isnull().sum()

In [69]:
df.isnull().sum()

CrimeDateTime      0
Description        0
Inside_Outside     0
Weapon             0
Gender             0
Age                0
Race               0
Old_District       0
Neighborhood       0
Latitude           0
Longitude          0
PremiseType        0
Total_Incidents    0
dtype: int64

#### CrimeDateTime

In [70]:
df.CrimeDateTime.dtype

dtype('O')

In [71]:
df['CrimeDateTime']=pd.to_datetime(df['CrimeDateTime'], infer_datetime_format=True, errors='coerce')

In [72]:
df['CrimeDateTime'].dtype

datetime64[ns, UTC]

In [73]:
df.isnull().sum()

CrimeDateTime      4
Description        0
Inside_Outside     0
Weapon             0
Gender             0
Age                0
Race               0
Old_District       0
Neighborhood       0
Latitude           0
Longitude          0
PremiseType        0
Total_Incidents    0
dtype: int64

In [74]:
df = df.dropna(subset=['CrimeDateTime'])

In [75]:
df.Age.isnull().sum()

0

In [76]:
df.isnull().sum()

CrimeDateTime      0
Description        0
Inside_Outside     0
Weapon             0
Gender             0
Age                0
Race               0
Old_District       0
Neighborhood       0
Latitude           0
Longitude          0
PremiseType        0
Total_Incidents    0
dtype: int64

#### Year

In [77]:
df['Year']=df['CrimeDateTime'].dt.year
df.Year.dtype

dtype('int32')

In [78]:
df.head()

CrimeDateTime        Description Inside_Outside  \
0 2023-04-15 11:30:00+00:00            LARCENY         Inside   
1 2023-04-15 21:50:00+00:00            LARCENY         Inside   
2 2023-04-15 17:00:00+00:00  LARCENY FROM AUTO         Inside   
3 2023-04-15 21:15:00+00:00            ROBBERY         Inside   
4 2023-04-11 09:25:00+00:00     COMMON ASSAULT        Outside   

             Weapon   Gender  Age                       Race Old_District  \
0           UNKNOWN   Female   56                      WHITE     NORTHERN   
1           UNKNOWN  UNKNOWN   38                    UNKNOWN    NORTHEAST   
2           UNKNOWN     Male   56  BLACK_OR_AFRICAN_AMERICAN    SOUTHWEST   
3           HANDGUN     Male   54                      WHITE    NORTHWEST   
4  PERSONAL_WEAPONS     Male   58                      WHITE      WESTERN   

              Neighborhood   Latitude  Longitude           PremiseType  \
0                  HARWOOD  39.318181 -76.609534                STREET   
1  PULASKI INDUSTRIAL AREA  39.310944 -76.530961           GAS STATION   
2              MOUNT HOLLY  39.313038 -76.674519                STREET   
3     CENTRAL PARK HEIGHTS  39.337523 -76.664090                STREET   
4                   MOSHER  39.296418 -76.661570  DRUG STORE / MED  BL   

   Total_Incidents  Year  
0                1  2023  
1                1  2023  
2                1  2023  
3                1  2023  
4                1  2023

#### Month

In [79]:
df['Month']=df['CrimeDateTime'].dt.month
df.head()

CrimeDateTime        Description Inside_Outside  \
0 2023-04-15 11:30:00+00:00            LARCENY         Inside   
1 2023-04-15 21:50:00+00:00            LARCENY         Inside   
2 2023-04-15 17:00:00+00:00  LARCENY FROM AUTO         Inside   
3 2023-04-15 21:15:00+00:00            ROBBERY         Inside   
4 2023-04-11 09:25:00+00:00     COMMON ASSAULT        Outside   

             Weapon   Gender  Age                       Race Old_District  \
0           UNKNOWN   Female   56                      WHITE     NORTHERN   
1           UNKNOWN  UNKNOWN   38                    UNKNOWN    NORTHEAST   
2           UNKNOWN     Male   56  BLACK_OR_AFRICAN_AMERICAN    SOUTHWEST   
3           HANDGUN     Male   54                      WHITE    NORTHWEST   
4  PERSONAL_WEAPONS     Male   58                      WHITE      WESTERN   

              Neighborhood   Latitude  Longitude           PremiseType  \
0                  HARWOOD  39.318181 -76.609534                STREET   
1  PULASKI INDUSTRIAL AREA  39.310944 -76.530961           GAS STATION   
2              MOUNT HOLLY  39.313038 -76.674519                STREET   
3     CENTRAL PARK HEIGHTS  39.337523 -76.664090                STREET   
4                   MOSHER  39.296418 -76.661570  DRUG STORE / MED  BL   

   Total_Incidents  Year  Month  
0                1  2023      4  
1                1  2023      4  
2                1  2023      4  
3                1  2023      4  
4                1  2023      4

In [80]:
df['Month']=df['CrimeDateTime'].dt.month
df.head()

CrimeDateTime        Description Inside_Outside  \
0 2023-04-15 11:30:00+00:00            LARCENY         Inside   
1 2023-04-15 21:50:00+00:00            LARCENY         Inside   
2 2023-04-15 17:00:00+00:00  LARCENY FROM AUTO         Inside   
3 2023-04-15 21:15:00+00:00            ROBBERY         Inside   
4 2023-04-11 09:25:00+00:00     COMMON ASSAULT        Outside   

             Weapon   Gender  Age                       Race Old_District  \
0           UNKNOWN   Female   56                      WHITE     NORTHERN   
1           UNKNOWN  UNKNOWN   38                    UNKNOWN    NORTHEAST   
2           UNKNOWN     Male   56  BLACK_OR_AFRICAN_AMERICAN    SOUTHWEST   
3           HANDGUN     Male   54                      WHITE    NORTHWEST   
4  PERSONAL_WEAPONS     Male   58                      WHITE      WESTERN   

              Neighborhood   Latitude  Longitude           PremiseType  \
0                  HARWOOD  39.318181 -76.609534                STREET   
1  PULASKI INDUSTRIAL AREA  39.310944 -76.530961           GAS STATION   
2              MOUNT HOLLY  39.313038 -76.674519                STREET   
3     CENTRAL PARK HEIGHTS  39.337523 -76.664090                STREET   
4                   MOSHER  39.296418 -76.661570  DRUG STORE / MED  BL   

   Total_Incidents  Year  Month  
0                1  2023      4  
1                1  2023      4  
2                1  2023      4  
3                1  2023      4  
4                1  2023      4

#### Day_of_Month

In [81]:
df['day_of_month']=df['CrimeDateTime'].dt.day
df.head()

CrimeDateTime        Description Inside_Outside  \
0 2023-04-15 11:30:00+00:00            LARCENY         Inside   
1 2023-04-15 21:50:00+00:00            LARCENY         Inside   
2 2023-04-15 17:00:00+00:00  LARCENY FROM AUTO         Inside   
3 2023-04-15 21:15:00+00:00            ROBBERY         Inside   
4 2023-04-11 09:25:00+00:00     COMMON ASSAULT        Outside   

             Weapon   Gender  Age                       Race Old_District  \
0           UNKNOWN   Female   56                      WHITE     NORTHERN   
1           UNKNOWN  UNKNOWN   38                    UNKNOWN    NORTHEAST   
2           UNKNOWN     Male   56  BLACK_OR_AFRICAN_AMERICAN    SOUTHWEST   
3           HANDGUN     Male   54                      WHITE    NORTHWEST   
4  PERSONAL_WEAPONS     Male   58                      WHITE      WESTERN   

              Neighborhood   Latitude  Longitude           PremiseType  \
0                  HARWOOD  39.318181 -76.609534                STREET   
1  PULASKI INDUSTRIAL AREA  39.310944 -76.530961           GAS STATION   
2              MOUNT HOLLY  39.313038 -76.674519                STREET   
3     CENTRAL PARK HEIGHTS  39.337523 -76.664090                STREET   
4                   MOSHER  39.296418 -76.661570  DRUG STORE / MED  BL   

   Total_Incidents  Year  Month  day_of_month  
0                1  2023      4            15  
1                1  2023      4            15  
2                1  2023      4            15  
3                1  2023      4            15  
4                1  2023      4            11

#### Day_of_Year

In [82]:
df['day_of_year'] = df['CrimeDateTime'].dt.dayofyear
df.head()

CrimeDateTime        Description Inside_Outside  \
0 2023-04-15 11:30:00+00:00            LARCENY         Inside   
1 2023-04-15 21:50:00+00:00            LARCENY         Inside   
2 2023-04-15 17:00:00+00:00  LARCENY FROM AUTO         Inside   
3 2023-04-15 21:15:00+00:00            ROBBERY         Inside   
4 2023-04-11 09:25:00+00:00     COMMON ASSAULT        Outside   

             Weapon   Gender  Age                       Race Old_District  \
0           UNKNOWN   Female   56                      WHITE     NORTHERN   
1           UNKNOWN  UNKNOWN   38                    UNKNOWN    NORTHEAST   
2           UNKNOWN     Male   56  BLACK_OR_AFRICAN_AMERICAN    SOUTHWEST   
3           HANDGUN     Male   54                      WHITE    NORTHWEST   
4  PERSONAL_WEAPONS     Male   58                      WHITE      WESTERN   

              Neighborhood   Latitude  Longitude           PremiseType  \
0                  HARWOOD  39.318181 -76.609534                STREET   
1  PULASKI INDUSTRIAL AREA  39.310944 -76.530961           GAS STATION   
2              MOUNT HOLLY  39.313038 -76.674519                STREET   
3     CENTRAL PARK HEIGHTS  39.337523 -76.664090                STREET   
4                   MOSHER  39.296418 -76.661570  DRUG STORE / MED  BL   

   Total_Incidents  Year  Month  day_of_month  day_of_year  
0                1  2023      4            15          105  
1                1  2023      4            15          105  
2                1  2023      4            15          105  
3                1  2023      4            15          105  
4                1  2023      4            11          101

#### Week_of_Year

In [83]:
df['week_of_year']=df['CrimeDateTime'].dt.isocalendar().week
df.head()

CrimeDateTime        Description Inside_Outside  \
0 2023-04-15 11:30:00+00:00            LARCENY         Inside   
1 2023-04-15 21:50:00+00:00            LARCENY         Inside   
2 2023-04-15 17:00:00+00:00  LARCENY FROM AUTO         Inside   
3 2023-04-15 21:15:00+00:00            ROBBERY         Inside   
4 2023-04-11 09:25:00+00:00     COMMON ASSAULT        Outside   

             Weapon   Gender  Age                       Race Old_District  \
0           UNKNOWN   Female   56                      WHITE     NORTHERN   
1           UNKNOWN  UNKNOWN   38                    UNKNOWN    NORTHEAST   
2           UNKNOWN     Male   56  BLACK_OR_AFRICAN_AMERICAN    SOUTHWEST   
3           HANDGUN     Male   54                      WHITE    NORTHWEST   
4  PERSONAL_WEAPONS     Male   58                      WHITE      WESTERN   

              Neighborhood   Latitude  Longitude           PremiseType  \
0                  HARWOOD  39.318181 -76.609534                STREET   
1  PULASKI INDUSTRIAL AREA  39.310944 -76.530961           GAS STATION   
2              MOUNT HOLLY  39.313038 -76.674519                STREET   
3     CENTRAL PARK HEIGHTS  39.337523 -76.664090                STREET   
4                   MOSHER  39.296418 -76.661570  DRUG STORE / MED  BL   

   Total_Incidents  Year  Month  day_of_month  day_of_year  week_of_year  
0                1  2023      4            15          105            15  
1                1  2023      4            15          105            15  
2                1  2023      4            15          105            15  
3                1  2023      4            15          105            15  
4                1  2023      4            11          101            15

#### Dayname_of_week

In [84]:
df['dayname_of_week']=df['CrimeDateTime'].dt.dayofweek
day_map={0: 'Monday', 1: 'Tuesday', 2: 'Wednesday', 3:'Thursday', 4: 'Friday', 5:'Saturday', 6:'Sunday'}
df['dayname_of_week']=df['dayname_of_week'].map(day_map)
df.head()

CrimeDateTime        Description Inside_Outside  \
0 2023-04-15 11:30:00+00:00            LARCENY         Inside   
1 2023-04-15 21:50:00+00:00            LARCENY         Inside   
2 2023-04-15 17:00:00+00:00  LARCENY FROM AUTO         Inside   
3 2023-04-15 21:15:00+00:00            ROBBERY         Inside   
4 2023-04-11 09:25:00+00:00     COMMON ASSAULT        Outside   

             Weapon   Gender  Age                       Race Old_District  \
0           UNKNOWN   Female   56                      WHITE     NORTHERN   
1           UNKNOWN  UNKNOWN   38                    UNKNOWN    NORTHEAST   
2           UNKNOWN     Male   56  BLACK_OR_AFRICAN_AMERICAN    SOUTHWEST   
3           HANDGUN     Male   54                      WHITE    NORTHWEST   
4  PERSONAL_WEAPONS     Male   58                      WHITE      WESTERN   

              Neighborhood   Latitude  Longitude           PremiseType  \
0                  HARWOOD  39.318181 -76.609534                STREET   
1  PULASKI INDUSTRIAL AREA  39.310944 -76.530961           GAS STATION   
2              MOUNT HOLLY  39.313038 -76.674519                STREET   
3     CENTRAL PARK HEIGHTS  39.337523 -76.664090                STREET   
4                   MOSHER  39.296418 -76.661570  DRUG STORE / MED  BL   

   Total_Incidents  Year  Month  day_of_month  day_of_year  week_of_year  \
0                1  2023      4            15          105            15   
1                1  2023      4            15          105            15   
2                1  2023      4            15          105            15   
3                1  2023      4            15          105            15   
4                1  2023      4            11          101            15   

  dayname_of_week  
0        Saturday  
1        Saturday  
2        Saturday  
3        Saturday  
4         Tuesday

#### Day_of_week

In [85]:
df['day_of_week']=df['CrimeDateTime'].dt.dayofweek
df.head()

CrimeDateTime        Description Inside_Outside  \
0 2023-04-15 11:30:00+00:00            LARCENY         Inside   
1 2023-04-15 21:50:00+00:00            LARCENY         Inside   
2 2023-04-15 17:00:00+00:00  LARCENY FROM AUTO         Inside   
3 2023-04-15 21:15:00+00:00            ROBBERY         Inside   
4 2023-04-11 09:25:00+00:00     COMMON ASSAULT        Outside   

             Weapon   Gender  Age                       Race Old_District  \
0           UNKNOWN   Female   56                      WHITE     NORTHERN   
1           UNKNOWN  UNKNOWN   38                    UNKNOWN    NORTHEAST   
2           UNKNOWN     Male   56  BLACK_OR_AFRICAN_AMERICAN    SOUTHWEST   
3           HANDGUN     Male   54                      WHITE    NORTHWEST   
4  PERSONAL_WEAPONS     Male   58                      WHITE      WESTERN   

              Neighborhood   Latitude  Longitude           PremiseType  \
0                  HARWOOD  39.318181 -76.609534                STREET   
1  PULASKI INDUSTRIAL AREA  39.310944 -76.530961           GAS STATION   
2              MOUNT HOLLY  39.313038 -76.674519                STREET   
3     CENTRAL PARK HEIGHTS  39.337523 -76.664090                STREET   
4                   MOSHER  39.296418 -76.661570  DRUG STORE / MED  BL   

   Total_Incidents  Year  Month  day_of_month  day_of_year  week_of_year  \
0                1  2023      4            15          105            15   
1                1  2023      4            15          105            15   
2                1  2023      4            15          105            15   
3                1  2023      4            15          105            15   
4                1  2023      4            11          101            15   

  dayname_of_week  day_of_week  
0        Saturday            5  
1        Saturday            5  
2        Saturday            5  
3        Saturday            5  
4         Tuesday            1

#### Hour

In [86]:
df['Hour']=df['CrimeDateTime'].dt.strftime('%H')
df.head()

CrimeDateTime        Description Inside_Outside  \
0 2023-04-15 11:30:00+00:00            LARCENY         Inside   
1 2023-04-15 21:50:00+00:00            LARCENY         Inside   
2 2023-04-15 17:00:00+00:00  LARCENY FROM AUTO         Inside   
3 2023-04-15 21:15:00+00:00            ROBBERY         Inside   
4 2023-04-11 09:25:00+00:00     COMMON ASSAULT        Outside   

             Weapon   Gender  Age                       Race Old_District  \
0           UNKNOWN   Female   56                      WHITE     NORTHERN   
1           UNKNOWN  UNKNOWN   38                    UNKNOWN    NORTHEAST   
2           UNKNOWN     Male   56  BLACK_OR_AFRICAN_AMERICAN    SOUTHWEST   
3           HANDGUN     Male   54                      WHITE    NORTHWEST   
4  PERSONAL_WEAPONS     Male   58                      WHITE      WESTERN   

              Neighborhood   Latitude  ...           PremiseType  \
0                  HARWOOD  39.318181  ...                STREET   
1  PULASKI INDUSTRIAL AREA  39.310944  ...           GAS STATION   
2              MOUNT HOLLY  39.313038  ...                STREET   
3     CENTRAL PARK HEIGHTS  39.337523  ...                STREET   
4                   MOSHER  39.296418  ...  DRUG STORE / MED  BL   

  Total_Incidents  Year  Month  day_of_month  day_of_year  week_of_year  \
0               1  2023      4            15          105            15   
1               1  2023      4            15          105            15   
2               1  2023      4            15          105            15   
3               1  2023      4            15          105            15   
4               1  2023      4            11          101            15   

   dayname_of_week day_of_week  Hour  
0         Saturday           5    11  
1         Saturday           5    21  
2         Saturday           5    17  
3         Saturday           5    21  
4          Tuesday           1    09  

[5 rows x 21 columns]

In [87]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 549793 entries, 0 to 603893
Data columns (total 21 columns):
 #   Column           Non-Null Count   Dtype              
---  ------           --------------   -----              
 0   CrimeDateTime    549793 non-null  datetime64[ns, UTC]
 1   Description      549793 non-null  object             
 2   Inside_Outside   549793 non-null  object             
 3   Weapon           549793 non-null  object             
 4   Gender           549793 non-null  object             
 5   Age              549793 non-null  int32              
 6   Race             549793 non-null  object             
 7   Old_District     549793 non-null  object             
 8   Neighborhood     549793 non-null  object             
 9   Latitude         549793 non-null  float64            
 10  Longitude        549793 non-null  float64            
 11  PremiseType      549793 non-null  object             
 12  Total_Incidents  549793 non-null  int64              
 13  Year

In [88]:
df.isnull().sum()

CrimeDateTime      0
Description        0
Inside_Outside     0
Weapon             0
Gender             0
Age                0
Race               0
Old_District       0
Neighborhood       0
Latitude           0
Longitude          0
PremiseType        0
Total_Incidents    0
Year               0
Month              0
day_of_month       0
day_of_year        0
week_of_year       0
dayname_of_week    0
day_of_week        0
Hour               0
dtype: int64

In [89]:
df.shape

(549793, 21)

In [90]:
clean_df=df[df['Year']>=2013]

In [91]:
clean_df.shape

(454293, 21)

In [93]:
filepath = "clean_crime_data.csv"
clean_df.to_csv(filepath, index=False)

In [94]:
clean_df.Year.unique()

array([2023, 2022, 2021, 2020, 2019, 2018, 2017, 2024, 2016, 2015, 2014,
       2013])

In [95]:
clean_df.isnull().sum()

CrimeDateTime      0
Description        0
Inside_Outside     0
Weapon             0
Gender             0
Age                0
Race               0
Old_District       0
Neighborhood       0
Latitude           0
Longitude          0
PremiseType        0
Total_Incidents    0
Year               0
Month              0
day_of_month       0
day_of_year        0
week_of_year       0
dayname_of_week    0
day_of_week        0
Hour               0
dtype: int64